internal/overlord/logstate/syslog.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	syslogParserSize = 4 * 1024
+
+	canonicalPrivEnterpriseNum = 28978
+)
+
+type syslogForwarder struct {
+	serviceName string
+	targetName  string
+	target      plan.LogTarget
+
+	iterator servicelog.Iterator
+	parser   *servicelog.Parser
+	stopCh   chan struct{}
+
+	buffer *syslogBuffer
+	client *syslogClient
+}
+
+func newSyslogForwarder(
+	serviceName, targetName string, target plan.LogTarget, iterator servicelog.Iterator,
+) (*syslogForwarder, error) {
+	client, err := newSyslogClient(target.Location)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer := newSyslogBuffer()
+	return &syslogForwarder{
+		serviceName: serviceName,
+		targetName:  targetName,
+		target:      target,
+		iterator:    iterator,
+		parser:      servicelog.NewParser(iterator, lokiParserSize),
+		stopCh:      make(chan struct{}, 1),
+		buffer:      buffer,
+		client:      client,
+	}, nil
+}
+
```

Comment:
Should be something like:

```
 func (f *syslogForwarder) Forward() {
 	for f.iterator.Next(f.stopCh) {
 		for f.parser.Next() {
 			f.buffer.WriteEntry(f.parser.Entry())
 		}
 		if !f.buffer.IsEmpty() {
 			f.flushBuffer()
 		}
 	}
 }
```

Reply:
iterator's Next needs to be called

Reply:
Sure, I just copied how it was done in `internal/daemon/api_logs.go`

Reply:
> ```
>  func (f *syslogForwarder) Forward() {
>  	for f.iterator.Next(f.stopCh) {
>  		for f.parser.Next() {
>  			f.buffer.WriteEntry(f.parser.Entry())
>  		}
>  		if !f.buffer.IsEmpty() {
>  			f.flushBuffer()
>  		}
>  	}
>  }
> ```

@hpidcock this doesn't work as written. `f.iterator.Next(f.stopCh)` never returns, so none of the log entries ever get written.

I can't see what was wrong with my way, it's done the same way as the logs API, and it was successfully able to flush logs through to Loki/syslog.